src/networking.c

```diff
@@ -1590,41 +1590,38 @@ int _writeToClient(client *c, ssize_t *nwritten) {
         return C_OK;
     }
 
-    if (c->bufpos > 0) {
-        *nwritten = connWrite(c->conn,c->buf+c->sentlen,c->bufpos-c->sentlen);
-        if (*nwritten <= 0) return C_ERR;
-        c->sentlen += *nwritten;
-
-        /* If the buffer was sent, set bufpos to zero to continue with
-         * the remainder of the reply. */
-        if ((int)c->sentlen == c->bufpos) {
-            c->bufpos = 0;
-            c->sentlen = 0;
-        }
-    } else {
-        if (listLength(c->reply) > 1) {
-            struct iovec iov[IOV_MAX];
+    int nbuf = listLength(c->reply);
+    if (nbuf > 0) {
+        if (c->bufpos > 0) ++nbuf; 
+        if (nbuf > 1) {
```

Comment:
- I don't think we need `nbuf`. we can use `IOV_MAX` (it is only 1024 entries and allocated on stack).
- If you agree with previous claim, then we can replace the two conditions with single `if(c->bufpos > 0)`
- Since the static buffer is rather small ~16KB, kernel has no problem write it in one sys-call, and much more (I tested multiple 1MBs and all written in one sys-call). Note that current implementation won't attach it to a big reply buffer that follows.
- Regarding TLS, since it is not optimized, I think we better avoid writev()
- Based on last two arguments, maybe we can writev something like that:
```
    if ((listLength(c->reply) > 0) && (connGetType(c->conn) == CONN_TYPE_SOCKET)) {
        return writeMultiBufToSocket(c, nwritten);
    }
```
Thank you

Reply:
We can use `IOV_MAX` as the length of `iov`, but I don't think we can simply use `if (c->bufpos > 0)` to decide whether to use `writev()` cuz the reply list may still have multiple nodes even if the static buffer is empty.

Reply:
I wrote it above implicitly... since the static buffer is rather small, in the common case it will be written entirely, together with the reply buffer that follows. That is, I tried to simplified the condition, assuming that this optimization has little effect over big buffers.

Reply:
>     if ((listLength(c->reply) > 0) && (connGetType(c->conn) == CONN_TYPE_SOCKET)) {
>         return writeMultiBufToSocket(c, nwritten);
>     }

    @panjf2000 I think this judgment is right.
    The only possible edge condition is bufpos==0, and reply length is 1,
    which is actually enough to use write, but it feels like it makes the code more complicated.
    BTW, @moticless I actually don't like `== CONN_TYPE_SOCKET`, but I feel like I can avoid it.

Reply:
So we should move the current apporach to a new function called `writeMultiBufToSocket()` and restore the old code of `_writeToClient()`?

Reply:
@panjf2000 Perhaps we can verify that TLS also delivers performance gains before deciding whether to revert to the old code.
I'm inclined to add the `writeMultiBufToSocket` method, the code is getting too big now.

Reply:
For a start, hope you agree with me that current implementation of `CT_TLS::writev()` is not adequate for big buffers and we must avoid memcpy.

One way to achieve common interface, is to modify `CT_TLS::writev()` to decide whether to use `writev()` or multiple `write()`s based on some threshold.

IMO,
If we have 16kb static buffer and 1MB reply buffer
- For CT_Socket, will be a single sys-call of `writev()` for both buffers (currently imp of `_writeToClient()` won't do it) . 
- For CT_TLS, will have two write()s (rather than memcpy and single sys-call writev().)

If we have two buffers of size 1MB, then 
- For CT_Socket, I think we better see two distinct `write()` - preserve previous behavior , unless we goanna put an effort to test performance (I don't think it is an interesting case to challenge and put an effort.)
- For CT_TLS,  will have two write()s

Reply:
@panjf2000 I benchmarked again in the morning, and I agree with @moticless that writev's merging of large packets does not result in a significant performance gain.

My test steps.
1. create a string key of 1m
2. use `redis-benchmark -P 100 -n 10000 get key` to test so that each clientReplyBlock ~1M, and c->reply length is 100

Reply:
Well, I guess I can set up a new function and move the new code into it, then restore the _writeToClient(), but I am concerned that there'll be duplicate code.